### PR TITLE
[JENKINS-58223] Add directive generator for KubernetesDeclarativeAgent

### PR DESCRIPTION
**Pull request by Vlatombe**
Fix inline helps

Upstream PRs
* [x] https://github.com/jenkinsci/workflow-cps-plugin/pull/362
* [x] https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/396

### CODE REVIEW

**Comment by jglick** [comment]
Does the configuration UI actually work? I remember looking into this a while back and finding that it was broken generally, even for agent impls built into `pipeline-model-definition` that did have `config` views. A screenshot would be nice here.

**Comment by Vlatombe** [comment]
Capture d’écran 2020-06-16 à 16 42 14

**Comment by jglick** [comment]
```groovy
podRetention never()
```

Is this not the default?

**Comment by Vlatombe** [comment]
There is a *default* setting which depends on the global plugin setting (which is *never*)

**Comment by Vlatombe** [comment]
> finding that it was broken generally, even for agent impls built into pipeline-model-definition that did have config views

https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/390, and also the new ones I referenced at the top of this PR.

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgent/help-customWorkspace.html [hunk 1]
```diff
@@ -0,0 +1 @@
+Allows to checkout the scm in a custom workspace directory.
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgent/help-customWorkspace.html:1]:
```suggestion
Allows the SCM repository to be checked out in a custom workspace directory.
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly [hunk 2]
```diff
@@ -22,29 +22,31 @@
...
 
-  <f:entry field="inheritFrom" title="${%Pod template to inherit from}"
-    help="/plugin/kubernetes/help/inheritFrom.html">
-    <f:textbox default="${descriptor.defaultInheritFrom}"/>
+  <f:entry field="inheritFrom" title="${%Pod template to inherit from}">
+    <f:select default="${descriptor.defaultInheritFrom}"/>
   </f:entry>
 
-  <f:entry title="${%Containers}" description="${%List of container in the agent pod}">
+  <f:entry field="containers" title="${%Containers}" description="${%List of containers in the agent pod}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly:29]:
```suggestion
  <f:entry field="containers" title="${%Containers}" description="${%Containers in the agent pod}">
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly [hunk 3]
```diff
@@ -22,29 +22,31 @@
...
       <f:repeatableHeteroProperty field="containers" hasHeader="true" addCaption="Add Container"
                                     deleteCaption="Delete Container" />
   </f:entry>
 
-  <f:entry title="${%Volumes}" description="${%List of volumes to mount in agent pod}">
+  <f:entry field="envVars"  title="${%EnvVars}" description="${%List of environment variables to set in all container of the pod}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly:34]:
```suggestion
  <f:entry field="envVars"  title="${%EnvVars}" description="${%Environment variables to set in all containers of the pod}">
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly [hunk 4]
```diff
@@ -22,29 +22,31 @@
...
+    <f:repeatableHeteroProperty field="envVars" hasHeader="true" addCaption="${%Add Environment Variable}"
+                                deleteCaption="${%Delete Environment Variable}" />
+  </f:entry>
+
+  <f:entry field="volumes"  title="${%Volumes}" description="${%List of volumes to mount in agent pod}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly:39]:
```suggestion
  <f:entry field="volumes"  title="${%Volumes}" description="${%Volumes to mount in agent pod}">
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly [hunk 5]
```diff
@@ -22,29 +22,31 @@
...
     <f:repeatableHeteroProperty field="volumes" hasHeader="true" addCaption="Add Volume"
                                 deleteCaption="Delete Volume" />
   </f:entry>
 
-  <f:entry title="${%Annotations}" description="${%List of annotations to add to the pod metadata}"
-    field="annotations" help="/plugin/kubernetes/help/annotations.html">
+  <f:entry field="annotations"  title="${%Annotations}" description="${%List of annotations to add to the pod metadata}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly:44]:
```suggestion
  <f:entry field="annotations"  title="${%Annotations}" description="${%Annotations to add to the pod metadata}">
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly [hunk 6]
```diff
@@ -22,29 +22,31 @@
...
     <f:repeatableHeteroProperty field="annotations" hasHeader="true" addCaption="Add Annotation"
                                 deleteCaption="Delete Annotation" />
   </f:entry>
 
-  <f:entry field="yaml" title="${%Raw yaml for the Pod}"
-    help="/plugin/kubernetes/help/yaml.html">
+  <f:entry field="yaml" title="${%Raw yaml for the Pod}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly:49]:
```suggestion
  <f:entry field="yaml" title="${%Raw YAML for the Pod}">
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly [hunk 7]
```diff
@@ -53,8 +55,7 @@
   </f:entry>
 
   <f:advanced>
-        <f:entry field="instanceCap" title="${%Max number of instances}"
-          help="/plugin/kubernetes/help/instanceCap.html">
+        <f:entry field="instanceCap" title="${%Max number of instances}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/config.jelly:58]:
```suggestion
        <f:entry field="instanceCap" title="${%Maximum number of instances}">
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/help-label.html [hunk 8]
```diff
@@ -2,5 +2,14 @@
     Jenkins node label to bind.
     If left blank, one will be generated for you,
     and inside the step it will be bound to the variable <code>POD_LABEL</code>
-    so you can use this as the argument to the <code>node</code> step.
+    so you can use this as the argument to the <code>node</code> step.<br/>
+
+    Example:
+    <pre>
+        podTemplate (...) {
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/help-label.html:9]:
```suggestion
        podTemplate(...) {
```

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/help.html [hunk 9]
```diff
@@ -1,3 +1,13 @@
...
-    Defines a container inside the kubernetes plugin configuration
-</div>
+<p>
+    Defines a Kubernetes pod template that can be used to create nodes.
+<p>
+<p>
+    Example:
+</p>
+<pre>
+podTemplate (...) {
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep/help.html:8]:
```suggestion
podTemplate(...) {
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgentUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Collections;
+import java.util.Map;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.isA;
+
+public class KubernetesDeclarativeAgentUnitTest {
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgentUnitTest.java:22]: Not really a unit test if using `JenkinsRule`.

Would `@ClassRule` be appropriate?

**Comment by jglick** [comment]
> There is a _default_ setting which depends on the global plugin setting (which is _never_)

OK…but if I do not change the value of this pulldown from what is initially presented, I expect to not see anything in Groovy.

**Comment by Vlatombe** [comment]
Yes, the default value of the dropdown is _Default_ and this produces an empty output in Groovy.

Capture d’écran 2020-06-16 à 17 47 23

File: pom.xml [hunk 11]
```diff
@@ -223,6 +233,12 @@
       <artifactId>test-harness</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>docker-workflow</artifactId>
```
[the remark this answers is not in the document]
Comment (Vlatombe) [reply pom.xml:238]: Needed to keep https://github.com/jenkinsci/kubernetes-plugin/blob/da02f75fbf975a7751af4a768ac571241d9fcc46/src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/docker.groovy#L16 working.

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate/config.jelly [hunk 12]
```diff
@@ -34,7 +34,7 @@
     <f:checkbox default="true"/>
   </f:entry>
 
-    <f:entry title="${%EnvVars}" description="${%List of environment variables to set in agent pod}">
+    <f:entry title="${%Environment Variables}" field="envVars" description="${%List of environment variables to set in agent pod}">
       <f:repeatableHeteroProperty field="envVars" hasHeader="true" addCaption="${%Add Environment Variable}"
```
[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/resources/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate/config.jelly:38]: https://github.com/jenkinsci/jenkins/pull/4807

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgentUnitTest.java (hunk 10 above)
Comment (Vlatombe) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgentUnitTest.java:22]: Yes, this is by opposition to `KubernetesDeclarativeAgentTest` which is performing builds, so need an actual Kubernetes environment. Can probably use `@ClassRule` here.

File: pom.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (jglick) [reply pom.xml:238]: Oh right.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java [hunk 13]
```diff
@@ -402,7 +402,7 @@ public DescriptorImpl() {
         @SuppressWarnings("unused") // by stapler/jelly
         public ListBoxModel doFillCloudItems() {
             ListBoxModel result = new ListBoxModel();
-            result.add("-- any --", "");
+            result.add("-any-", "");
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java:405]: FWIW I suggested
```suggestion
            result.add("—any—", "");
```
or whatever looks nice in your browser

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java [hunk 14]
```diff
@@ -413,8 +413,8 @@ public ListBoxModel doFillCloudItems() {
         public ListBoxModel doFillInheritFromItems(@QueryParameter("cloud") String cloudName) {
             cloudName = Util.fixEmpty(cloudName);
             ListBoxModel result = new ListBoxModel();
-            result.add("-- Default inheritance --", "<default>");
-            result.add("-- Disable inheritance --", " ");
+            result.add("-Default inheritance-", "<default>");
+            result.add("-Disable inheritance-", " ");
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java:417]: similarly
```suggestion
            result.add("—Default inheritance—", "<default>");
            result.add("—Disable inheritance—", " ");
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java (hunk 13 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java:405]: Ah,I didn't see that it was a different character.

Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java]: https://getitwriteonline.com/articles/en-dashes-em-dashes/ Didn't even know the _en dash_.

Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java]: And let us not forget U+2015 HORIZONTAL BAR ― which crops up a lot in Irish novels IIRC!

File: pom.xml [hunk 15]
```diff
@@ -49,7 +49,8 @@
     <java.level>8</java.level>
     <jenkins.version>2.222.4</jenkins.version>
     <no-test-jar>false</no-test-jar>
-    <pipeline-model-definition.version>1.6.0</pipeline-model-definition.version>
+    <pipeline-model-definition.version>1.7.1-rc1836.8776418d532d</pipeline-model-definition.version><!-- https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/396 -->
+    <workflow-cps.version>2.81-rc2492.83316f41dee4</workflow-cps.version><!-- https://github.com/jenkinsci/workflow-cps-plugin/pull/362 -->
```
Comment (Vlatombe) [inline pom.xml:53]:
```suggestion
    <workflow-cps.version>2.81</workflow-cps.version>
```

File: pom.xml [hunk 16]
```diff
@@ -49,7 +49,8 @@
     <java.level>8</java.level>
     <jenkins.version>2.222.4</jenkins.version>
     <no-test-jar>false</no-test-jar>
-    <pipeline-model-definition.version>1.6.0</pipeline-model-definition.version>
+    <pipeline-model-definition.version>1.7.1-rc1836.8776418d532d</pipeline-model-definition.version><!-- https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/396 -->
```
Comment (Vlatombe) [inline pom.xml:52]:
```suggestion
    <pipeline-model-definition.version>1.7.2</pipeline-model-definition.version>
```

**Comment by jglick** [comment]
@Vlatombe why the empty `help-nodeProperties.html`?

**Comment by Vlatombe** [comment]
Probably an oversight.

**Pull request merged**